pkg/apis/build/v1alpha1/target_path_validation.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1alpha1
+
+import (
+	"strings"
+
+	"github.com/knative/pkg/apis"
+)
+
+// Node is constructed to track targetPath structure in source
+// Value tracks the value of node. Parent holds the address of parent nodes
+type node struct {
+	parent *node
+	value  string
+}
+
+// insert function breaks the path and inserts every element into nodeMap
+// insert function throws error if the path is already traversed by any
+// previous targetPaths
+func insertNode(path string, nodeMap map[string]*node) *apis.FieldError {
```

Comment:
Why does this one function need its own file?

Can we get a TestInsertNode unit test in `target_path_validation_test.go`?

I'm not 100% clear this works correctly (seems like it will only allow a single source to have a github.com prefix? Which is definitely not what we want. We just want to make sure that A is not a subset of B and B is not a subset of A, but it is perfectly fine to have A and B both be subsets of C so long as C is not a source, aka

OK:
A: github.com/foo/bar
B: github.com/foo/frobber

Not ok:
A: github.com/foo
B: github.com/foo/bar

Not ok:
A: github.com/foo/bar
B: github.com/foo

I think this method is trying to raise an error if any two paths start with github.com/foo, right?

Reply:
True. I missed some use cases. Let me fix it and ping you on this thread

Reply:
@fejta : I have updated the PR to include the above test cases.